File: Partie1.py
# ================================
# Analyse pollution - Réseau Métro IDF
# ================================

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn.model_selection import train_test_split
from sklearn.cluster import KMeans
from sklearn.preprocessing import LabelEncoder
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import classification_report, confusion_matrix, accuracy_score

# ------------------------------
# 1. Nettoyage du dataset
# ------------------------------
df = pd.read_csv("qualite-de-lair-dans-le-reseau-de-transport-francilien.csv", sep=";")

# Supprimer lignes vides
df.dropna(how="all", inplace=True)

# Supprimer colonnes inutiles
colonnes_a_supprimer = [
    "Lien vers les mesures en direct", "Durée des mesures",
    "Mesures d’amélioration mises en place ou prévues",
    "point_geo", "pollution_air", "air", "niveau", "actions", "niveau_pollution"
]
df.drop(columns=[col for col in colonnes_a_supprimer if col in df.columns], inplace=True)

# Nettoyage des valeurs textuelles
df.replace("pas de données", pd.NA, inplace=True)

# Supprimer doublons
df.drop_duplicates(inplace=True)

# Supprimer lignes sans pollution mesurée
df.dropna(subset=["Niveau de pollution aux particules", "Niveau de pollution"], how="all", inplace=True)

# ------------------------------
# 2. ACP - Est-ce pertinent ?
# ------------------------------
"""
Non pertinent directement, car :
- La plupart des colonnes sont qualitatives ou textuelles
- Peu de variables numériques exploitables pour ACP
Une ACP nécessiterait d’encoder de nombreuses colonnes et de compléter beaucoup de valeurs manquantes.
"""

# ------------------------------
# 3. Filtrer stations métro et splitter
# ------------------------------
df_metro = df[df["Nom de la ligne"].str.contains("Métro", na=False)]

train, test = train_test_split(df_metro, test_size=0.3, random_state=42)
train.to_csv("train.csv", index=False)
test.to_csv("test.csv", index=False)

# ------------------------------
# 4. Modèle K-Means clustering
# ------------------------------
df_geo = df_metro[["stop_lat", "stop_lon"]].dropna()

kmeans = KMeans(n_clusters=3, random_state=42)
df_geo["cluster"] = kmeans.fit_predict(df_geo)

# Visualisation des clusters
plt.figure(figsize=(8,6))
sns.scatterplot(data=df_geo, x="stop_lon", y="stop_lat", hue="cluster", palette="Set2")
plt.title("Clustering des stations de métro (K-Means)")
plt.xlabel("Longitude")
plt.ylabel("Latitude")
plt.savefig("clusters_kmeans.png")
plt.close()

# ------------------------------
# 5. Modèle KNN
# ------------------------------
df_knn = df_metro[["stop_lat", "stop_lon", "Niveau de pollution"]].dropna()

# Encodage de la cible
le = LabelEncoder()
df_knn["pollution_encoded"] = le.fit_transform(df_knn["Niveau de pollution"])

X = df_knn[["stop_lat", "stop_lon"]]
y = df_knn["pollution_encoded"]

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

knn = KNeighborsClassifier(n_neighbors=3)
knn.fit(X_train, y_train)
y_pred = knn.predict(X_test)

# ------------------------------
# 6. Évaluation des modèles
# ------------------------------
print("=== Évaluation du modèle KNN ===")
print("Accuracy:", accuracy_score(y_test, y_pred))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred))
print("Classification Report:\n", classification_report(y_test, y_pred, target_names=le.classes_))

# ------------------------------
# 7. Dashboard (à développer)
# ------------------------------
# Configuration
st.set_page_config(page_title="Pollution Métro IDF", layout="wide")

# Titre principal
st.title("🚇 Dashboard Pollution - Réseau Métro Île-de-France")

# Charger les données
@st.cache_data
def load_data():
    df = pd.read_csv("metro_pollution_qgis.csv")
    df = df[df["Nom de la ligne"].str.contains("Métro", na=False)]
    df.replace("pas de données", pd.NA, inplace=True)
    return df.dropna(subset=["Niveau de pollution"])

df = load_data()

# Filtre latéral
lignes = df["Nom de la ligne"].dropna().unique()
ligne_selection = st.sidebar.multiselect("Sélectionner une ou plusieurs lignes de métro :", lignes, default=lignes)

df_filtered = df[df["Nom de la ligne"].isin(ligne_selection)]

# KPIs
st.subheader("📊 Indicateurs clés")
col1, col2, col3 = st.columns(3)
col1.metric("Nombre de stations", len(df_filtered))
col2.metric("Niveaux de pollution mesurés", df_filtered["Niveau de pollution"].nunique())
col3.metric("Lignes sélectionnées", len(ligne_selection))

# Carte interactive (si données géo)
st.subheader("🗺️ Carte des stations sélectionnées")
st.map(df_filtered.rename(columns={"stop_lat": "lat", "stop_lon": "lon"}))

# Graphiques
st.subheader("📉 Répartition des niveaux de pollution")
fig, ax = plt.subplots(figsize=(8,4))
sns.countplot(data=df_filtered, x="Niveau de pollution", order=df_filtered["Niveau de pollution"].value_counts().index, ax=ax)
st.pyplot(fig)

st.subheader("📈 Pollution par ligne de métro")
fig2, ax2 = plt.subplots(figsize=(10,5))
sns.countplot(data=df_filtered, y="Nom de la ligne", hue="Niveau de pollution", ax=ax2)
st.pyplot(fig2)

# Footer
st.markdown("---")
st.caption("Données : Île-de-France Mobilités • Visualisation : Streamlit • Projet : Qualité de l'air métro")

File: Partie1.R
# ==============================
# Analyse pollution métro RATP
# ==============================

# Chargement des packages
library(dplyr)
library(readr)
library(ggplot2)
library(caret)
library(class)
library(cluster)
set.seed(123)

# 1. Nettoyage du dataset
df <- read_csv2("qualite-de-lair-dans-le-reseau-de-transport-francilien.csv")

# Supprimer lignes vides
df <- df[rowSums(is.na(df)) != ncol(df), ]

# Colonnes inutiles à supprimer
colonnes_a_supprimer <- c(
  "Lien vers les mesures en direct", 
  "Durée des mesures", 
  "Mesures d’amélioration mises en place ou prévues", 
  "point_geo", "pollution_air", "air", "niveau", "actions", "niveau_pollution"
)
df <- df %>% select(-one_of(intersect(colnames(df), colonnes_a_supprimer)))

# Remplacer "pas de données" par NA
df[df == "pas de données"] <- NA

# Supprimer doublons
df <- distinct(df)

# Supprimer lignes sans pollution mesurée
df <- df %>% filter(!(is.na(`Niveau de pollution aux particules`) & is.na(`Niveau de pollution`)))

# ==========================================
# 2. ACP : Est-elle intéressante ?
# ==========================================
# Non, car la majorité des variables sont qualitatives.
# Pour faire une ACP pertinente, il faut :
# - Des variables numériques continues
# - Peu de valeurs manquantes
# Donc une ACP n’est pas directement applicable sauf après encodage très complet et nettoyage avancé.

# ==========================================
# 3. Filtrer métro et créer train/test
# ==========================================
df_metro <- df %>% filter(grepl("Métro", `Nom de la ligne`))

# Sauvegarde train/test
train_idx <- sample(1:nrow(df_metro), 0.7 * nrow(df_metro))
train <- df_metro[train_idx, ]
test <- df_metro[-train_idx, ]

write_csv(train, "train.csv")
write_csv(test, "test.csv")

# ==========================================
# 4. K-means pour regrouper les stations
# ==========================================
# Encodage simplifié : on ne garde que les colonnes géographiques
df_metro_geo <- df_metro %>% select(stop_lat, stop_lon) %>% drop_na()

# K-means clustering (3 groupes arbitrairement)
kmeans_model <- kmeans(df_metro_geo, centers = 3, nstart = 10)

# Ajouter les clusters au dataset
df_metro_geo$cluster <- as.factor(kmeans_model$cluster)
ggplot(df_metro_geo, aes(x = stop_lon, y = stop_lat, color = cluster)) +
  geom_point() +
  labs(title = "Clustering des stations métro par K-means")

# ==========================================
# 5. KNN : prédire niveau de pollution
# ==========================================
# Préparation : transformer variable à prédire en facteur
df_knn <- df_metro %>% 
  filter(!is.na(`Niveau de pollution`)) %>%
  select(stop_lat, stop_lon, `Niveau de pollution`)

df_knn$`Niveau de pollution` <- as.factor(df_knn$`Niveau de pollution`)

# Split
idx <- sample(1:nrow(df_knn), 0.7 * nrow(df_knn))
train_knn <- df_knn[idx, ]
test_knn <- df_knn[-idx, ]

# KNN
pred_knn <- knn(train = train_knn[, c("stop_lat", "stop_lon")],
                test = test_knn[, c("stop_lat", "stop_lon")],
                cl = train_knn$`Niveau de pollution`,
                k = 3)

# ==========================================
# 6. Évaluation des modèles
# ==========================================
# Évaluation KNN
confusionMatrix(pred_knn, test_knn$`Niveau de pollution`)

# Évaluation K-means : pas une prédiction, mais un regroupement
# => Peut être visualisé et comparé à des zones géographiques
